Mathlib/NumberTheory/ModularForms/JacobiTheta/TwoVariable.lean

```diff
@@ -14,119 +15,501 @@ This file defines the two-variable Jacobi theta function
+  rw [(by push_cast; ring : (π * I * n ^ 2 * 2 : ℂ) = (n ^ 2 :) * (2 * π * I)), exp_int_mul,
+    exp_two_pi_mul_I, one_zpow]
+
+lemma jacobiTheta₂'_add_left (z τ : ℂ) : jacobiTheta₂' (z + 1) τ = jacobiTheta₂' z τ := by
+  unfold jacobiTheta₂' jacobiTheta₂'_term jacobiTheta₂_term
+  congr 1 with n
+  simp_rw [mul_add, Complex.exp_add, mul_one, mul_comm _ (n : ℂ), exp_int_mul_two_pi_mul_I, mul_one]
+
+lemma jacobiTheta₂'_add_left' (z τ : ℂ) :
+    jacobiTheta₂' (z + τ) τ = cexp (-π * I * (τ + 2 * z)) * (jacobiTheta₂' z τ -
+      2 * π * I * jacobiTheta₂ z τ) := by
+  rcases le_or_lt τ.im 0 with hτ | hτ
+  · simp_rw [jacobiTheta₂_undef _ hτ, jacobiTheta₂'_undef _ hτ, mul_zero, sub_zero, mul_zero]
+  have (n : ℤ) : jacobiTheta₂'_term n (z + τ) τ =
+      cexp (-π * I * (τ + 2 * z)) * (jacobiTheta₂'_term (n + 1) z τ -
+      2 * π * I * jacobiTheta₂_term (n + 1) z τ):= by
+    simp only [jacobiTheta₂'_term, jacobiTheta₂_term]
+    conv_rhs => rw [← sub_mul, mul_comm, mul_assoc, ← Complex.exp_add, Int.cast_add, Int.cast_one,
+      mul_add, mul_one, add_sub_cancel_right]
+    congr 2
+    ring
+  rw [jacobiTheta₂', funext this, tsum_mul_left, ← (Equiv.subRight (1 : ℤ)).tsum_eq]
+  simp_rw [Equiv.subRight_apply, sub_add_cancel, tsum_sub (hasSum_jacobiTheta₂'_term z hτ).summable
+    ((hasSum_jacobiTheta₂_term z hτ).summable.mul_left _), tsum_mul_left]
+  rfl
+
+lemma jacobiTheta₂'_neg_left (z τ : ℂ) : jacobiTheta₂' (-z) τ = -jacobiTheta₂' z τ := by
+  rw [jacobiTheta₂', jacobiTheta₂', ← tsum_neg, ← (Equiv.neg ℤ).tsum_eq]
+  congr 1 with n
+  simp only [jacobiTheta₂'_term, jacobiTheta₂_term]
+  rw [Equiv.neg_apply, ← neg_mul]
+  push_cast
+  ring_nf
+
+lemma jacobiTheta₂'_conj (z τ : ℂ) :
+    conj (jacobiTheta₂' z τ) = jacobiTheta₂' (conj z) (-conj τ) := by
+  rw [← neg_inj, ← jacobiTheta₂'_neg_left, jacobiTheta₂', jacobiTheta₂', conj_tsum, ← tsum_neg]
+  congr 1 with n
+  simp_rw [jacobiTheta₂'_term, jacobiTheta₂_term, map_mul, ← Complex.exp_conj, map_add, map_mul,
+    ← ofReal_int_cast,← ofReal_ofNat, map_pow, conj_ofReal, conj_I]
+  ring_nf
+
+/-!
+## Functional equations
+-/
+
 /-- The functional equation for the Jacobi theta function: `jacobiTheta₂ x τ` is an explict factor
 times `jacobiTheta₂ (x / τ) (-1 / τ)`. This is the key lemma behind the proof of the functional
-equation for Dirichlet L-series. -/
-theorem jacobiTheta₂_functional_equation (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
-    jacobiTheta₂ z τ =
-      1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (-π * I * z ^ 2 / τ) * jacobiTheta₂ (z / τ) (-1 / τ) := by
+equation for L-series of even Dirichlet characters. -/
+theorem jacobiTheta₂_functional_equation (z τ : ℂ) : jacobiTheta₂ z τ =
+    1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (-π * I * z ^ 2 / τ) * jacobiTheta₂ (z / τ) (-1 / τ) := by
+  rcases le_or_lt (im τ) 0 with hτ | hτ
+  · rw [jacobiTheta₂_undef z hτ, jacobiTheta₂_undef, mul_zero]
+    rw [neg_div, neg_im, one_div, inv_im, neg_nonpos]
```

Comment:
Is there a reason why there are two `rw`s instead of one?
(I see that combining them would likely not save a line, so it's OK either way.)

Reply:
They're not working on the same goal. The original goal is closed by `mul_zero`; but `jacobiTheta₂_undef` has spawned a new side goal of proving `(-1 / τ).im ≤ 0` which the last two lines deal with.